core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,420 @@
+            } else {
+                // look through the back, starting from the end, for added tombstones
+                final Tombstone lastAddedTombstone = previousTombstones.get(previousTombstones.size() - 1);
+                for (int i = currentTombstones.size() - 1; i >= 0; i--) {
+                    final Tombstone tombstone = currentTombstones.get(i);
+                    if (lastAddedTombstone.equals(tombstone)) {
+                        // already have this entry, so will have all previous entries as well
+                        break;
+                    } else {
+                        added.add(tombstone);
+                    }
+                }
+                // look through the front, starting at the beginning, for removed tombstones
+                final Tombstone firstTombstone = currentTombstones.get(0);
+                final int previousSize = previousTombstones.size();
+                for (int i = 0; i < previousSize; i++) {
+                    final Tombstone tombstone = previousTombstones.get(i);
+                    if (firstTombstone.equals(tombstone)) {
+                        // current list has this entry, so no more removing
+                        break;
+                    } else {
+                        removed.add(tombstone);
+                    }
+                }
+            }
+            Collections.reverse(added); // added list was created from back to front, so have to reverse
+            this.added = Collections.unmodifiableList(added);
+            this.removed = Collections.unmodifiableList(removed);
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            out.writeList(added);
+            out.writeList(removed);
+        }
+
+        @Override
+        public IndexGraveyard apply(final IndexGraveyard old) {
+            final List<Tombstone> oldQueue = old.tombstones;
+            final Deque<Tombstone> newQueue = new ArrayDeque<>(oldQueue);
+            for (Tombstone tombstone : removed) {
+                newQueue.remove(tombstone);
+            }
+            for (Tombstone tombstone : added) {
+                newQueue.addLast(tombstone);
+            }
+            return new IndexGraveyard.Builder(newQueue).build();
+        }
+
+        /** The index tombstones (deletions) that were added between two states */
+        public List<Tombstone> getAdded() {
+            return added;
+        }
+
+        /** The index tombstones (deletions) that were removed between two states */
+        public List<Tombstone> getRemoved() {
+            return removed;
+        }
+    }
+
+    /**
+     * An individual tombstone entry for representing a deleted index.
+     */
+    final public static class Tombstone implements ToXContent, StreamInputReader<Tombstone>, Writeable<Tombstone> {
+
+        public static final Tombstone PROTOTYPE = new Tombstone(new Index("_na_", "_na_"), 0L);
```

Comment:
I can here, but I will have to add a prototype to the `IndexGraveyard` because I am now implementing it as a `MetaData.Custom`